### PR TITLE
Price Oracle

### PR DESCRIPTION
**Pull request by shaopengw**
To do:

- [x] Fix existing unit tests.
- [x] Add unit tests for `Arrays` lib, `PriceOracleDataSource`,  and `SinglePriceOracle`.

### CODE REVIEW

File: contracts/impls/SimplePriceOracle.sol [hunk 1]
```diff
@@ -4,47 +4,93 @@ import "../interfaces/PriceOracleInterface.sol";
...
     }
 
-    function getPrice(address addr) external view returns (uint) {
-        return prices[addr];
+    function feedPrice(address key, uint price) external onlyPriceFeeder {
+        priceRecords[key][msg.sender] = PriceOracleStructs.PriceRecord(price, block.timestamp);
+        hasUpdate[key] = true;
+    }
+
+    function findMedianPrice(address key, uint expireIn) public returns (uint) {
```
Comment (xlc) [inline contracts/impls/SimplePriceOracle.sol]: needs to be protected

File: contracts/impls/SimplePriceOracle.sol [hunk 2]
```diff
@@ -4,47 +4,93 @@ import "../interfaces/PriceOracleInterface.sol";
...
+            PriceOracleStructs.PriceRecord storage record = priceRecords[key][priceFeeders[i]];
+            if (record.timestamp > expireAt) {
+                validPrices.push(record.price);
+            }
+        }
+
+        return Arrays.findMedian(validPrices);
+    }
+
+    function setHasUpdate(address key, bool value) public {
```
Comment (xlc) [inline contracts/impls/SimplePriceOracle.sol]: needs to be protected

File: contracts/impls/SimplePriceOracle.sol [hunk 3]
```diff
@@ -4,47 +4,93 @@ import "../interfaces/PriceOracleInterface.sol";
...
-    bool public constant isPriceOracle = true;
+contract PriceOracleDataSource is PriceFeederRole {
+    // key => feeder => price record
+    mapping(address => mapping(address => PriceOracleStructs.PriceRecord)) private priceRecords;
+    // key => hasUpdate
+    mapping(address => bool) public hasUpdate;
 
-    event PriceUpdated(address indexed addr, uint price);
+    // to store temp non-expired records for `findMedianPrice` method
+    uint[] private validPrices;
```
Comment (xlc) [inline contracts/impls/SimplePriceOracle.sol]: is it possible to remove this and use memory array instead?

File: contracts/libs/Arrays.sol [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+        }
+
+        // put partitioning item v at arr[j]
+        swap(arr, low, j);
+
+        return j;
+    }
+
+    function swap(uint[] storage arr, uint i, uint j) private {
+        uint temp = arr[i];
```
Comment (xlc) [inline contracts/libs/Arrays.sol]: does `(arr[i], arr[j]) = (arr[j], arr[i])` work?

File: contracts/impls/SimplePriceOracle.sol (hunk 3 above)
Comment (shaopengw) [reply contracts/impls/SimplePriceOracle.sol]: It is possible, but as memory arrays are fix sized, to get a filtered result array for finding median, extra copy operations and number of filtered items variable would be needed. I guess the cost (complexity and gas) is worthy?

**Comment by shaopengw** [comment]
A quick question: should we make oracle's `getPrice` and `setPrice` functions protected?

**Comment by xlc** [comment]
`getPrice` can be public. `setPrice` should be private with `feedPrice` protected with price feeder only

**Pull request merged**